### ISSUE TITLE
Provide access to working dir and environment from project settings without using settings helper files

### ISSUE THREAD

**Issue by francoechea** [csoltenborn/GoogleTestAdapter#175]
Visual studio allows you to set up environment variables just before you debug your executable via the <LocalDebuggerEnvironment> variable. However, since GTA does not take the settings directly from the VS project (e.g., $(WorkDir), etc) if any tests rely on setting of environment variables they will fail.

Is there a way within GTA to set the Enviroment?

**Comment by csoltenborn** [comment]
I haven't really understood the issue, I must admit - can you provide an example or elaborate a bit on your problem?

**Comment by LukaszMendakiewicz** [comment]
I believe this is about reading Project > Properties > Debugging > Environment and using it during test run. Here's how we do that in TAfBT currently: https://github.com/Microsoft/vs-boost-unit-test-adapter/blob/6a1275f0c496fac398640a8fe46c756110deef25/VisualStudioAdapter/VisualStudio.cs#L63

**Comment by csoltenborn** [comment]
Ok, I see - I'm thus tagging this enhancement.

**Comment by francoechea** [comment]
Yes, precisely: being able to set up environment variables while debugging just as in Visual Studio

**Comment by csoltenborn** [comment]
For the record: #241 asked for the working directory of the project settings.

**Comment by Khraekoh** [comment]
Could you also add support for the debugger command line arguments? I would think this is very similar to using the debugger environment variables, which is why I added a comment here.

We use the command line to set additional options for our tests such as adding extra logging and telling a test where data files can be found.

**Comment by csoltenborn** [comment]
For the record: #255 asks for the`ExecutablePath`.

**Comment by csoltenborn** [comment]
Version 0.16.0 of GTA introduced support for [settings helper files](https://github.com/csoltenborn/GoogleTestAdapter#settings-helper-files) which can be used to realize all requirements mentioned in this issue! Check it out...

However, I will still keep this issue opened - it would of course be desirable to have a slightly more convenient way to get that information.

**Comment by dagophil** [comment]
@csoltenborn Can you explain how I can use a settings helper file to set an environment variable that can be accessed from the code?

I have set Debugging > Environment to MYENVVAR=hello. When running the tests with F5 the variable is set correctly. However, when running the tests from the test explorer the value is not set.

**Comment by csoltenborn** [comment]
I believe that I already did exactly that ;-) - have you read the docs linked above? Please be more specific with your question...

**Comment by dagophil** [comment]
I have a test that looks like this:

```
#include <gtest/gtest.h>
#include <cstdlib>

TEST(gtestenv, find_env_variable)
{
    char const* const var = std::getenv("MYENVVAR");
    ASSERT_TRUE(var != nullptr);
    EXPECT_EQ(std::string(var), "hello");
}
```
In the project settings, the Debugging Environment is set to `MYENVVAR=hello`. When running my test executable with F5 (`Start Debugging`) , the test succeeds. However, when running from the test explorer, the test fails because `var` is `nullptr`, which means that the environment variable is not set.

I manually created a `foo_tests.exe.gta_settings_helper` next to the output file `foo_tests.exe`. It has the following content:
```
LocalDebuggerEnvironment=MYENVVAR=hello
```
However, the tests still fails when run from the test explorer. I thought that it may be a problem with the `=`, so I added quotes, but the test still fails:
```
LocalDebuggerEnvironment="MYENVVAR=hello"
```

I tried to use a post build event to generate the `.gta_settings_helper` file, because that's what the [settings helper file documentation](https://github.com/csoltenborn/GoogleTestAdapter#settings-helper-files) said.
```
echo LocalDebuggerEnvironment=$(LocalDebuggerEnvironment) > $(TargetPath).gta_settings_helper
```
Now the build fails with the following output:
```
1>foo_tests.vcxproj -> D:\Code\src\gtest_env_test\build_vs2017\Debug\foo_tests.exe
1>LocalDebuggerEnvironment=MYENVVAR=hello
1>Syntaxfehler.
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(138,5): error MSB3073: The command "echo LocalDebuggerEnvironment=MYENVVAR=hello
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(138,5): error MSB3073:  > D:\Code\src\gtest_env_test\build_vs2017\Debug\foo_tests.exe.gta_settings_helper
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(138,5): error MSB3073: :VCEnd" exited with code 255.
1>Done building project "foo_tests.vcxproj" -- FAILED.
```
The following post build event works, so I think I got the syntax right:
```
echo SolutionPath=$(SolutionPath) > $(TargetPath).gta_settings_helper
```
**TLDR:**
My question is still the same as the initial comment in this thread https://github.com/csoltenborn/GoogleTestAdapter/issues/175#issue-268710110: Is there a way within GTA to set the enviroment (variables)?

**Comment by csoltenborn** [comment]
Ok - from you mentioning that it works with F5, but not with GTA, I had assumed that you hadn't read the docs, because they clearly state that this is expected.

There is no (direct) way to influence the Environment the test executables are executed in. The settings helper files' task is a different one: It makes all VS macros available for GTA. The idea is to generate a helper file as you have done; the key-value pair (in your case: key=`LocalDebuggerEnviroment`, value=`MYENVVAR=hello` can then be used within GTA settings, i.e., the placeholder `$(LocalDebuggerEnvironment)` can be used within GTA's settings, and will be replaced with `MYENVVAR=hello` at test discovery and test execution time.

However, this does not appear to be what you need. I have two suggestions which might solve your problem:
* Switch from using env variables to command line parameters, which can be influenced by GTA. You would need to make three changes:
  * Change test code to parsing  from the passed parameters instead of using `std::getenv()`
  * Pass `MYENVVAR` as a parameter at the test project setup (Project -> Properties -> Debugging -> Arguments) to enable execution via F5 (why is this needed anyways?)
  * Configure GTA such that `MYENVVAR` and its value are passed to the test executable (option [*Additional test execution parameters*](https://github.com/csoltenborn/GoogleTestAdapter/blob/db6f65b2a645da48a78fcdead715e86e92e747a2/SampleTests/SampleTests.gta.runsettings#L6) - use settings helper file as described above if value depends on some VS macros to be resolved)
* Use test setup/teardown scripts to configure your enviroment

The "Syntaxfehler" above is out of scope of GTA, but my feeling is that you might not even need settings helper files...

**Comment by dagophil** [comment]
Thank you for the quick and detailed reply. I thought this issue was about making environment variables available to the test code, but that was a misunderstanding on my side and I apologize.

Refactoring the code so that is does not rely on environment variables or std::getenv would be a working solution. However, the code that uses the environment variable is buried inside a lib and our users currently rely on the software being customizable via environment variables. I dont want to change this off-handedly.

I tried to use your second suggestion: I created a batch file with the content `SET MYENVVAR=hello` and added it under _Test setup batch file_. However, the test is still not passing. Is it possible that the test runner somehow resets the environment variables? I confirmed that the batch file is called correctly by letting it create a dummy file.

**Comment by csoltenborn** [comment]
Don't worry - this issue has ended up as a mixture of all kinds of stuff...

GTA does not change the environment despite extending the `PATH` variable if the according option is set. I must admit that env variables are not exactly my area of expertise, but I have one more idea: Are using `setx` as decribed [here](https://stackoverflow.com/questions/3803581/setting-a-system-environment-variable-from-a-windows-batch-file)?

**Comment by dagophil** [comment]
Using `setx` instead of `set` works. However, this changes the value of the environment variable permanently. It's okay as a workaround though.

**Comment by csoltenborn** [comment]
That's what I expected... You might consider scripting to save the current value of the variables in the setup script, and reversing your changes in the teardown script using the saved values.

**Comment by csoltenborn** [comment]
@dagophil FYI: I have today released GTA 0.18.0 which allows to change environment variables without affecting the system's env variables permanently (i.e., you can probably get rid of the workaround of setting the vars by means of the setup script).

**Comment by dagophil** [comment]
@csoltenborn Together with a post build event that creates a settings helper file, I was able to expand our custom VS macros and pass them as environment variables to the tests. This is really great :) Thank you!

**Comment by csoltenborn** [comment]
[This branch](/csoltenborn/GoogleTestAdapter/tree/Storage/%23175_VS_project_properties) contains a proof-of-concept implementation for receiving project properties from VS. However, I decided to not finish implementation for two reasons:
* Although slightly more convenient, the feature does not add any new functionality (helper files allow to make use of project properties already)
* In contrast to helper files, this approach does not work for NuGet packages and `VsTest.Console.exe`

I'm thus closing this issue, but keeping the branch as reference in case.
